File: Java/src/_0863_All_Nodes_Distance_K_in_Binary_Tree/Solution.java
package _0863_All_Nodes_Distance_K_in_Binary_Tree;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * Date: 2021/7/28 9:27
 * Content:
 */
public class Solution {

    HashMap<TreeNode, List<TreeNode>> g = new HashMap<>();
    List<Integer> ans = new ArrayList<>();

    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        dfs1(root);
        dfs2(target, null, k);
        return ans;
    }

    private void dfs1(TreeNode root) {

        if (!g.containsKey(root)) g.put(root, new ArrayList<>());

        if (root.left != null) {
            if (!g.containsKey(root.left)) g.put(root.left, new ArrayList<>());
            g.get(root).add(root.left);
            g.get(root.left).add(root);
            dfs1(root.left);
        }
        if (root.right != null) {
            if (!g.containsKey(root.right)) g.put(root.right, new ArrayList<>());
            g.get(root).add(root.right);
            g.get(root.right).add(root);
            dfs1(root.right);
        }
    }

    private void dfs2(TreeNode root, TreeNode father, int k) {
        if (k == 0) ans.add(root.val);
        else {
            for (TreeNode son : g.get(root))
                if (son != null && son != father)
                    dfs2(son, root, k - 1);
        }
    }

    public static void main(String[] args) {

        int nu = Integer.MIN_VALUE;  // 用Integer.MIN_VALUE表示二叉树 null
        int[] nums = {
                3,
                5, 1,
                6, 2, 0, 8,
                nu, nu, 7, 4
        };
        TreeNode root = (new MyTree(nums)).getRoot();
        TreeNode target = root.left;
        System.out.println((new Solution()).distanceK(root, target, 2));  // [1, 7, 4]
    }
}

File: Python/_0863_All_Nodes_Distance_K_in_Binary_Tree/Solution.py
# coding=utf-8
# Date: 2021/7/28 9:37
from MyTree import *


# 执行用时：28 ms, 在所有 Python3 提交中击败了99.72%的用户
# 内存消耗：15.2 MB, 在所有 Python3 提交中击败了74.51%的用户
class Solution:
    def __init__(self):
        self.g = {}
        self.ans = []

    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        self.dfs1(root)
        self.dfs2(target, None, k)
        return self.ans

    def dfs1(self, root):
        if root not in self.g:
            self.g[root] = []

        if root.left is not None:
            if root.left not in self.g:
                self.g[root.left] = []
            self.g[root].append(root.left)
            self.g[root.left].append(root)
            self.dfs1(root.left)
        if root.right is not None:
            if root.right not in self.g:
                self.g[root.right] = []
            self.g[root].append(root.right)
            self.g[root.right].append(root)
            self.dfs1(root.right)

    def dfs2(self, root, father, k):
        if k == 0:
            self.ans.append(root.val)
        else:
            for son in self.g[root]:
                if son is not None and son != father:
                    self.dfs2(son, root, k - 1)


if __name__ == "__main__":
    nu = None
    nums = [
        3,
        5, 1,
        6, 2, 0, 8,
        nu, nu, 7, 4
    ]
    root = MyTree(nums).getRoot()
    target = root.left
    print(Solution().distanceK(root, target, 2))  # [1, 7, 4]
